learn/writing-secure-ballerina-code.md

```diff
@@ -76,60 +76,51 @@ Security-sensitive functions and remote methods of Ballerina standard libraries
-   var dt = testDB->select("SELECT NAME FROM STUDENT WHERE ID = " + studentId,
-                           ResultStudent);
-   testDB.stop();
+    jdbc:Client|sql:Error jdbcClient = new("jdbc:mysql://localhost:3306/testdb", "test", "test");
+    if (jdbcClient is jdbc:Client) {
+        // Construct student ID based on user input.
+        string studentId = "S_" + args[0];
+
+        // Execute select query using the untrusted (tainted) student ID
+        stream<record{}, error> resultStream = jdbcClient->query("SELECT NAME FROM STUDENT WHERE ID = " + studentId);
+        error? status = jdbcClient.close();
+    }
 }
 ```
 
 The Ballerina compiler will generate an error:
 
 ```
-tainted value passed to sensitive parameter 'sqlQuery'
+tainted value passed to untainted parameter 'sqlQuery'
 ```
 
 In order to compile, the program is modified to use query parameters:
 
 ```ballerina
-jdbc:Parameter paramId = {sqlType:jdbc:TYPE_VARCHAR, value:studentId};
-var dt = testDB->select("SELECT NAME FROM STUDENT WHERE ID = ?", ResultStudent,
-                        paramId);
+stream<record{}, error> resultStream = jdbcClient->query(`SELECT NAME FROM STUDENT WHERE ID = ${<@untainted> studentId}`);
 ```
 
 Command-line arguments passed to Ballerina programs and inputs received through service resources are considered as tainted. Additionally, return values of certain functions are marked with the `@tainted` annotation to denote that the resulting value should be considered as untrusted data.
 
-For example, the `select` remote method of the `java:jdbc` client highlighted above returns a `@tainted table<record {}>|Error`. This means that any value read from a database is considered as untrusted.
+For example, the `query` remote method of the `java.jdbc` client highlighted above returns a `@tainted stream <record {}, sql:Error>`. This means that any value read from a database is considered as untrusted.
 
-When the Ballerina compiler can determine that a function is returning tainted data without tainted data being passed in as parameters to that function, it is required to annotate the function's return type as `@tainted`. If not, the function author has to clean up the data before returning. For instance, if you are to read from the database and return that result, you either need to annotate that function's return type as `@tainted` or you have to clean up and make sure the returned data is not tainted.
+When the Ballerina compiler can determine that a function is returning tainted data without tainted data being passed in as parameters to that function, it is required to annotate the function's return type `@tainted`. If not, the function author has to clean up the data before returning. For instance, if you are to read from the database and return the result, you either need to annotate that function's return type as `@tainted` or you have to clean up and make sure the returned data is not tainted.
```

Comment:
```suggestion
When the Ballerina compiler can determine that a function is returning tainted data without tainted data being passed in as parameters to that function, it is required to annotate the function's return type as `@tainted`. If not, the function author has to clean up the data before returning. For instance, if you are to read from the database and return the result, you either need to annotate that function's return type as `@tainted` or you have to clean up and make sure the returned data is not tainted.
```

Reply:
@praneesha would it be ok to use `with` instead of `as` here?

Reply:
+1